File: leetcode_2583/KthLargestSumInABinaryTree.py
from collections import deque
import heapq
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left: Optional['TreeNode'] = None, right: Optional['TreeNode'] = None):
        self.val = val
        self.left = left
        self.right = right

class KthLargestSumInABinaryTree:
    def kthLargestSum(self, root: TreeNode, k: int) -> int:
        q = deque([root])
        min_heap = []
        
        while q:
            level_sum = 0
            for i in range(len(q)):
                node = q.popleft()
                level_sum += node.val
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)
            heapq.heappush(min_heap, level_sum)
            if len(min_heap) > k:
                heapq.heappop(min_heap)
        return -1 if len(min_heap) < k else min_heap[0]
        
def main():
    main_instance = KthLargestSumInABinaryTree()
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.left = TreeNode(3)
    print(main_instance.kthLargestSum(root, 1))

if __name__ == "__main__":
    main()

File: leetcode_2583/KthLargestSumInABinaryTree.java
import java.util.*;

class KthLargestSumInABinaryTree{
    public long kthLargestSum(TreeNode root, int k){
        List<Long> res = new ArrayList<>(); // To store sum of each level
        Queue<TreeNode> q = new LinkedList<>(); // Queue for level order traversal
        q.add(root);  // Start BFS from the root node

        while(!q.isEmpty())
        {
            int n = q.size(); // Number of nodes at current level
            long sum = 0; // Sum of node values at the current level

            for(int i=0; i<n; i++)
            {
                TreeNode node = q.poll();
                sum += node.val;

                if(node.left != null) q.add(node.left);
                if(node.right != null) q.add(node.right);
            }
            res.add(sum); // Store the sum of the current level
        }

        if(k > res.size()) return -1;

        res.sort(Collections.reverseOrder()); // Sort level sums in descending order
        return res.get(k-1); // Return the kth largest sum
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.left = new TreeNode(3);

        KthLargestSumInABinaryTree obj = new KthLargestSumInABinaryTree();
        System.out.println(obj.kthLargestSum(root, 1)); 
    }


}